File: framework-test/src/main/java/com/example/framework/testsupport/BaseCase.java
package com.example.framework.testsupport;

import com.fasterxml.jackson.databind.ObjectMapper;
import jakarta.validation.ConstraintViolation;
import jakarta.validation.Validation;
import jakarta.validation.Validator;
import lombok.extern.slf4j.Slf4j;
import org.hibernate.validator.messageinterpolation.ParameterMessageInterpolator;

import java.util.Collection;
import java.util.Locale;
import java.util.Set;
import java.util.function.Supplier;

import static org.assertj.core.api.Assertions.assertThat;

/**
 * @author MaGuangZu
 * @since 2024-05-15
 */
@Slf4j
public class BaseCase {

	protected ObjectMapper objectMapper = new ObjectMapper();

	private Supplier<String> acceptLanguage = null;

	private final Validator validator = Validation.byDefaultProvider()
		.configure()
		.messageInterpolator(
			new ParameterMessageInterpolator(
				Set.of(
					Locale.CHINA,
					Locale.US
				),  // supported locales
				Locale.CHINA,  // default locale
				context -> {
					if (acceptLanguage != null) {
						var acceptedLanguages = Locale.LanguageRange.parse(acceptLanguage.get());
						var resolvedLocales = Locale.filter(acceptedLanguages, context.getSupportedLocales());
						if (!resolvedLocales.isEmpty()) {
							return resolvedLocales.getFirst();
						}
					}
					return context.getDefaultLocale();
				},
				false
			)
		)
		.buildValidatorFactory()
		.getValidator();

	protected <T> Set<ConstraintViolation<T>> validate(T obj) {
		acceptLanguage = null;
		return validator.validate(obj);
	}

	<T> Set<ConstraintViolation<T>> validate(Locale lang, T obj) {
		acceptLanguage = lang::getLanguage;
		return validator.validate(obj);
	}

	protected <T> void validateResult(Collection<String> collection, Set<ConstraintViolation<T>> validateResult) {
		validateResult.forEach(it -> assertThat(collection).contains(it.getMessage()));
	}

	protected <T> void printResult(Set<ConstraintViolation<T>> validateResult) {
		if (!validateResult.isEmpty()) {
			validateResult.forEach(System.out::println);
		}
	}
}

File: framework-test/src/main/kotlin/com/example/framework/testsupport/BaseCase.kt
package com.example.framework.testsupport

import com.fasterxml.jackson.databind.ObjectMapper
import jakarta.validation.ConstraintViolation
import jakarta.validation.Validation
import jakarta.validation.Validator
import org.assertj.core.api.Assertions.assertThat
import org.hibernate.validator.messageinterpolation.ParameterMessageInterpolator
import org.hibernate.validator.spi.messageinterpolation.LocaleResolverContext
import java.util.*
import java.util.Locale.LanguageRange
import java.util.function.Supplier

/**
 * @author MaGuangZu
 * @since 2021-09-26
 */
open class BaseCase {

	val objectMapper = ObjectMapper()

	private var acceptLanguage: Supplier<String>? = null

	private var validator: Validator = Validation.byDefaultProvider()
		.configure()
		.messageInterpolator(
			ParameterMessageInterpolator(
				mutableSetOf(
					Locale.CHINA,
					Locale.US
				),  // supported locales
				Locale.CHINA,  // default locale
				{ context: LocaleResolverContext ->
					if (acceptLanguage != null) {
						val acceptedLanguages = LanguageRange.parse(acceptLanguage!!.get())
						val resolvedLocales = Locale.filter(acceptedLanguages, context.supportedLocales)
						if (resolvedLocales.isNotEmpty()) {
							return@ParameterMessageInterpolator resolvedLocales[0]
						}
					}
					return@ParameterMessageInterpolator context.defaultLocale
				}, false
			)
		)
		.buildValidatorFactory()
		.validator

	fun <T> validate(obj: T): Set<ConstraintViolation<T>> {
		acceptLanguage = null
		return validator.validate(obj)
	}

	fun <T> validate(lang: Locale, obj: T): Set<ConstraintViolation<T>> {
		acceptLanguage = Supplier { lang.language }
		return validator.validate(obj)
	}

	fun <T> validateResult(collection: Collection<String>, validateResult: Set<ConstraintViolation<T>>) {
		validateResult.stream().forEach {
			assertThat(collection).contains(it.message)
		}
	}

	fun <T> printResult(validateResult: Set<ConstraintViolation<T>>) {
		if (validateResult.isNotEmpty()) {
			validateResult.forEach { println(it) }
		}
	}

}
